### PR TITLE
Fix Search - Arrow navigation on search stops working

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Details

[Comment](https://github.com/Expensify/App/issues/6131#issuecomment-955619193)
### Fixed Issues

$ https://github.com/Expensify/App/issues/6131

### Tests / QA

1. Press `Control + K`
2. Enter something search
3. Press the down arrow several times
4. Verify that you can navigate the search list with the arrow keys

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/29673073/141044840-0c16a176-83b0-4789-be01-b69192355790.mp4

#### Mobile Web

Tested for crash.
#### Desktop

https://user-images.githubusercontent.com/29673073/141047135-a65616da-976e-428e-8bb2-5f287ae1b8c9.mov
#### iOS

Tested for crash.
#### Android

Tested for crash.

### CODE REVIEW

**Review by Julesssss — DISMISSED** [review]
Looks good, tests well.

**Comment by Julesssss** [comment]
Holding, see [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1636397689083200) for more info.

File: src/components/OptionsList/index.native.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    <BaseOptionsList
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...props}
+        ref={ref}
+
+        // Both `keyboardDismissMode` & `onScrollBeginDrag` props are needed to ensure that virtual keyboard is
+        // dismissed on all platforms.
+        // eslint-disable-next-line react/jsx-props-no-multi-spaces
+        keyboardDismissMode="on-drag"
+        onScrollBeginDrag={() => Keyboard.dismiss()}
```
Comment (parasharrajat) [inline src/components/OptionsList/index.native.js:15]: OK. This is fine. But as we are now abstracting away these props for native and web. Then it means we can get rid of the one which is not needed for native. You will get more info in the original PR that added this. I don't remember atm.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
There is one issue here apart from all good. it will remove the keyboard hiding from the Mobile web. 
May be you can test that and make it so that it does not break M-web.

**Comment by parasharrajat** [comment]
DO NOT MERGE yet. I have a few comments above...Thanks.

**Comment by parasharrajat** [comment]
@rushatgabhane Do you mind attaching videos for M-web and native? It affects that behaviour directly so its worth attaching here.

**Comment by parasharrajat** [comment]
@rushatgabhane Any update?

**Comment by rushatgabhane** [comment]
Oops, looks like I missed the notification earlier.
@parasharrajat I'll update you within 2 hours.

**Comment by rushatgabhane** [comment]
Ah okay I get it, pull down to dismiss keyboard doesn't work on **mWeb**.
Things work fine on Native because there are no changes to it.

https://user-images.githubusercontent.com/29673073/143293031-b3382913-34c7-47cd-a6a0-1fa15db6b701.mp4

**Comment by rushatgabhane** [comment]
Native works fine.

https://user-images.githubusercontent.com/29673073/143293704-0dea3aa9-3902-4bcb-be8b-0aa7abfd4e86.mp4

File: src/components/OptionsList/index.native.js (hunk 1 above)
Comment (rushatgabhane) [reply src/components/OptionsList/index.native.js:15]: Got it! We can get rid of `keyboardDismissMode="on-drag"`

[Reference commit](https://github.com/Expensify/App/commit/c72eb4add194cdee88261ab290da6de38cc98a51#diff-6a3f4819485fb409e83240f58a2b8ee3c096ad40718d8f9e12929c1adb888a38R238)

**Comment by rushatgabhane** [comment]
TLDR; While my proposal fixes the issue,
It causes a regression: keyboard won't dismiss on scrolling down for **mWeb**.

https://github.com/Expensify/App/pull/6272#issuecomment-978117669

**Comment by parasharrajat** [comment]
I am sure we can fix that.

**Comment by rushatgabhane** [comment]
Well, we need to use `keyboardDismissMode="on-drag" ` for mWeb only.
Do you think checking for mWeb and conditionally using the prop is a good idea?

We could do that using a regex test for mWeb.
Or we could assume smallScreenWidth as mWeb.

What do you say?

**Comment by parasharrajat** [comment]
In our app, We consider isSmallScreenWidth to be a Mobile device and thus it can be used for MWeb. I know that a tablet could be Mweb too so I am confused about this as well. Let's take an opinion here @Julesssss Could you please help?

I also think subscribing to Onyx for this component is expensive.

**Comment by Julesssss** [comment]
I'm tempted to say that we shouldn't care that pull-to-refresh is missing for mWeb. I could be wrong, but I believe that `isSmallScreenWidth` can be true for small native mobile devices. So if the trade-off is between losing the keyboard dismiss behavior on mWeb or native, I would rather keep it working on native. Does anyone disagree or have a better solution?

**Comment by rushatgabhane** [comment]
Arrow navigation on web, and keyboard dismiss on native are more important than keyboard dismiss on mWeb.
The tradeoff is justified imo.

**Comment by parasharrajat** [comment]
Yeah. If we use `isSmallScreenWidth` then we can keep the mobile web behaviour as well. My main concern here is that It should create performance issues. `OptionList` is the most costly component of the app.
@rushatgabhane Could you please test it as well?

**Comment by rushatgabhane** [comment]
> I believe that isSmallScreenWidth can be true for small native mobile devices

@Julesssss I don't think that'd be a problem because `isSmallScreenWidth`  won't be used in `index.native.js`

**Comment by rushatgabhane** [comment]
@parasharrajat 
I don't know how to benchmark, or what metrics to look for in this case. 

But I couldn't find any glaring performance issues, and I think we're good to merge.
I hope this is what you were looking for.

Apoligies for the forced push, had to remove an unverified commit.

**Comment by parasharrajat** [comment]
Thanks I will review shortly

**Comment by rushatgabhane** [comment]
Also, I doubt withWindowDimensions is using Onyx.
It's just a Higher order component which adds isSmallScreenWidth as a prop.

So I don't think we're adding any substantial overhead.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
@rushatgabhane Please merge main into this PR. There have been many important changes recently.

File: src/components/OptionsList/BaseOptionsList.js [hunk 2]
```diff
@@ -235,12 +137,8 @@ class OptionsList extends Component {
...
                     keyboardShouldPersistTaps="always"
-
-                    // Both `keyboardDismissMode` & `onScrollBeginDrag` props are needed to ensure that virtual keyboard is
-                    // dismissed on all platforms.
-                    // eslint-disable-next-line react/jsx-props-no-multi-spaces
-                    keyboardDismissMode="on-drag"
-                    onScrollBeginDrag={() => Keyboard.dismiss()}
+                    keyboardDismissMode={this.props.keyboardDismissMode}
+                    onScrollBeginDrag={this.props.onScrollBeginDrag}
                     contentContainerStyle={[...this.props.contentContainerStyles]}
```
Comment (parasharrajat) [inline src/components/OptionsList/BaseOptionsList.js:142]: Could you please make it so
```suggestion
                    contentContainerStyle={this.props.contentContainerStyles}
```

File: src/components/OptionsList/baseOptionsListPropTypes.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    disableRowInteractivity: PropTypes.bool,
+
+    /** Callback to execute when the SectionList lays out */
+    onLayout: PropTypes.func,
+
+    /** Determines whether the keyboard gets dismissed in response to a drag */
+    keyboardDismissMode: PropTypes.string,
+
+    /** Called when the user begins to drag the scroll view */
+    onScrollBeginDrag: PropTypes.func,
```
Comment (parasharrajat) [inline src/components/OptionsList/baseOptionsListPropTypes.js:84]: These two are not passed to OptionList Component. so let's make a change so prop definitions are correct.
Update the props for `BaseOptionsList` like
```
const propTypes = {
    ...OptionListPropTypes,
     /** Determines whether the keyboard gets dismissed in response to a drag */
    keyboardDismissMode: PropTypes.string,

    /** Called when the user begins to drag the scroll view */
    onScrollBeginDrag: PropTypes.func,
}
```
same for defaultprops.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
Please remove HOLD from the title as well..

File: src/components/OptionsList/baseOptionsListPropTypes.js (hunk 3 above)
Comment (rushatgabhane) [reply src/components/OptionsList/baseOptionsListPropTypes.js]: Done

File: src/components/OptionsList/index.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import withWindowDimensions from '../withWindowDimensions';
+import {propTypes, defaultProps} from './baseOptionsListPropTypes';
+
+const OptionsList = forwardRef((props, ref) => (
+    <BaseOptionsList
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...props}
+        ref={ref}
+
+        // eslint-disable-next-line react/jsx-props-no-multi-spaces
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/components/OptionsList/index.js]: nope. nice catch

**Review by parasharrajat — APPROVED** [review]
LGTM, Thanks for the changes. 
All yours @Julesssss

**Review by Julesssss — APPROVED** [review]
LGTM
